### PR TITLE
Add binary exponentiation, response to issue #1734

### PR DESCRIPTION
**Pull request by masongreen30**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1886"></a>

### CODE REVIEW

File: divide_and_conquer/binary_exponentiation.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * (https://cp-algorithms.com/algebra/binary-exp.html)
+ * @details
+ * Given a base and an exponent, we want to return the resulting value.
+ * Normal exponentiation calculates a^n in O(n), but using binary
+ * exponentiation we are able to calculate a^n in O(logn).
+ * @author [Mason Green](https://github.com/masongreen30)
+ */
+
+#include <cassert>
+#include <iostream>
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:13]: Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * @namespace divide_and_conquer
+ * @brief Divide and Conquer algorithms
+ */
+namespace divide_and_conquer {
+/**
+ * @namespace karatsuba_algorithm
+ * @brief Function for Binary Exponentiation
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:22]:
```suggestion
 * @brief Function for the [Binary Exponentiation](https://cp-algorithms.com/algebra/binary-exp.html) algorithm implementation
```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        } 
+        calculated *= calculated;
+    }
+    return result;
+}
+}  // namespace binary_exponentiation
+}  // namespace divide_and_conquer
+
+/**
+ * @brief Function to test binary_exponentiation
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:53]:
```suggestion
 * @brief Self-test implementations
```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    std::cout << "successful" << std::endl;
+}
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    // Test algorithm
+    test();
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:88]:
```suggestion
    test();  // run self-test implementations
```

File: divide_and_conquer/binary_exponentiation.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    return 0;
+}
```
Comment (Panquesito7) [inline divide_and_conquer/binary_exponentiation.cpp:90]: Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/masongreen30/C-Plus-Plus/actions

**Comment by masongreen30** [comment]
I believe it's enabled now, but I'm not entirely sure.

**Pull request closed without merging**